src/smk/layer/layer-vector.js

```diff
@@ -21,54 +21,82 @@ include.module( 'layer.layer-vector-js', [ 'layer.layer-js' ], function () {
         var cv = $( '<canvas width="' + width * mult + '" height="' + height + '">' ).get( 0 )
         var ctx = cv.getContext( '2d' )
 
-        var styles = [].concat( self.config.style )
+        const legendData = [];
+        legendData.push({
+            title: self.config.legend.title || self.config.title,
+            style: self.config.style
+        });
+
+        if (self.config.conditionalStyles) {
+            self.config.conditionalStyles.forEach(cs => {
+                const defaultStyle = self.config.style ? self.config.style : {};
+                self.config.conditionalStyles.forEach(conditionalStyle => {
+                    conditionalStyle.conditions.forEach(condition => {
+                        const combinedStyle = Object.assign({}, defaultStyle);
+                        Object.assign(combinedStyle, condition.style);
+                        legendData.push({
+                            title: condition.label || condition.value,
+                            style: combinedStyle,
+                            indent: true
+                        });
+                    });
+                });
+            });
+        }
 
         return SMK.UTIL.resolved( 0 )
             .then( drawPoint )
             .then( drawLine )
             .then( drawFill )
             .then( function () {
-                return [ {
-                    url: cv.toDataURL( 'image/png' ),
-                    title: self.config.legend.title || self.config.title
-                } ]
+                return legendData;
             } )
 
         function drawPoint( offset ) {
             if ( !self.config.legend.point ) return offset 
 
-            return SMK.UTIL.makePromise( function ( res, rej ) {
-                if ( styles[ 0 ].markerUrl ) {
-                    var img = $( '<img>' )
-                        .on( 'load', function () {
-                            var r = img.width / img.height
-                            if ( r > 1 ) r = 1 / r
-                            ctx.drawImage( img, offset, 0, height * r, height )
-                            res( offset + width )
-                        } )
-                        .on( 'error', res )
-                        .attr( 'src', viewer.resolveAttachmentUrl( styles[ 0 ].markerUrl, null, 'png' ) )
-                        .get( 0 )
-                }
-                else {
+            if (legendData.length === 1 && legendData[0].style.markerUrl) {
+                loadMarkerImage(offset)
+                .then((offset) => {
+                    legendData[0].url = cv.toDataURL( 'image/png' );
+                    return offset;
```

Comment:
Do you know what the purpose of returning `offset` is? Is it necessary?

Reply:
I don't believe it's used. We do need to have a return value in promises though, so I left it all as-is for consistency, and the chance this is used in some other context.

Reply:
Is the return value from a promise a requirement of the SMK framework? It's not required by JavaScript.

Reply:
I might be misinterpreting, but I got the impression a return is required from reading reference like https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises - "Important: Always return results, otherwise callbacks won't catch the result of a previous promise (with arrow functions, () => x is short for () => { return x; }). If the previous handler started a promise but did not return it, there's no way to track its settlement anymore, and the promise is said to be "floating".".